File: problems/023-storage-evacuation/solutions/python/main.py
import sys

it = iter(sys.stdin.buffer.read().split())
n = int(next(it))
C = int(next(it))
A = int(next(it))
R = int(next(it))
v = []
for _ in range(n):
    v.append((int(next(it)), int(next(it)), int(next(it)), next(it), next(it)))
need = max(0, sum(x[0] for x in v) - C, R - A)
if need > sum(x[0] for x in v):
    print("IMPOSSIBLE")
else:
    v.sort(key=lambda x: (x[1], x[2], x[3], x[4]))
    freed = k = 0
    while freed < need:
        freed += v[k][0]
        k += 1
    print(k, freed)
    for x in v[:k]:
        print(x[3].decode(), x[4].decode())

File: problems/023-storage-evacuation/solutions/typescript/main.ts
import * as std from "std";
type I = { z: bigint; p: number; u: bigint; a: string; k: string };
const t = std.in.readAsString().trim().split(/\s+/);
let q = 0,
  n = +t[q++],
  C = BigInt(t[q++]),
  A = BigInt(t[q++]),
  R = BigInt(t[q++]),
  T = 0n;
const v: I[] = [];
for (let i = 0; i < n; i++) {
  const x: I = {
    z: BigInt(t[q++]),
    p: +t[q++],
    u: BigInt(t[q++]),
    a: t[q++],
    k: t[q++],
  };
  T += x.z;
  v.push(x);
}
let need = T - C;
if (need < 0n) need = 0n;
if (R - A > need) need = R - A;
if (need > T) std.out.puts("IMPOSSIBLE\n");
else {
  v.sort((x, y) =>
    x.p - y.p || (x.u < y.u ? -1 : x.u > y.u ? 1 : 0) ||
    (x.a < y.a ? -1 : x.a > y.a ? 1 : 0) || (x.k < y.k ? -1 : x.k > y.k ? 1 : 0)
  );
  let f = 0n, k = 0;
  while (f < need) f += v[k++].z;
  const out: string[] = [`${k} ${f}`];
  for (let i = 0; i < k; i++) out.push(`${v[i].a} ${v[i].k}`);
  std.out.puts(out.join("\n") + "\n");
}
